library(Seurat)
"*******************************************************************************
DAY 1 Lecture 3/4
-----------------
After having completed this chapter you will be able to:

  - Load single cell data into R
  - Explain the basic structure of a Seurat object and extract count data and metadata
  - Calculate and visualize quality measures based on:
    * mitochondrial genes
    * ribosomal genes
    * hemoglobin genes
    * relative gene expression
  Interpret the above quality measures per cell.
  Perform cell filtering based on quality thresholds
*******************************************************************************"

# First, we will load a file specifying the different samples, 
# and create an object specifying the location of the count data:
sample_info <- read.csv("course_data/sample_info_course.csv") # a DataFrame
datadirs <- file.path("course_data", 
                      "count_matrices", 
                      sample_info$SampleName,
                      "outs",
                      "filtered_feature_bc_matrix") # vectorized file path list
# create named vector (dict in python) for datadirs
# beforehand, gsub (global substitution) the SampleNames
names(datadirs) <- gsub("_", "-", sample_info$SampleName) 
datadirs <- datadirs[1:3] # take the first three 

# This directory is part of the output generated by cellranger. 
# To load this data into R and generate a sparse matrix, run the following command:
sparse_matrix <- Seurat::Read10X(data.dir = datadirs)

# Have a look at the counts of the first 30 cells of three genes by running:
sparse_matrix[c("PECAM1", "CD8A", "TSPAN1"), 1:30]
# You will see many dots (zeros) and a few integers representing the counts per gene per cell.

# To generate a Seurat object, we will run CreateSeuratObject.
seu <- Seurat::CreateSeuratObject(counts = sparse_matrix,
                                  project = "pbmmc",
                                  min.cells = 3, # filter out genes present less than 3 cells
                                  min.features = 100) #filter out cells with less than 100 genes
# The object contains multi-level slots and lists
# You can get the information inside a slot with @

# return the a data frame with information on each cell
seu@meta.data
#Slots can be filled with other R objects, like lists, vectors, data frames or any other class. 

"-------------------------------------------------------------------------------
In addition to the original count table,
the Seurat object can therefore store a lot of information 
that is generated during your analysis, 
like results of a 
  - normalization (@assays$RNA@data) 
  - a PCA or UMAP (@reductions) 
  - and the clustering (@graphs).
-------------------------------------------------------------------------------"

# seu@assays$RNA@layers$counts is equivalent to GetAssayData(object = seu, layer = "counts")

table(seu@active.ident)
# returns the number of cells per sample.

head(seu@meta.data)
"-------------------------------------------------------------------------------
Giving you the names of three columns and a row for each cell:

  - orig_ident: the original identity (origin) of a cell.
  - nCount_RNA: the number of reads assigned to a cell.
  - nFeature_RNA: the number of expressed features (genes) per cell.
-------------------------------------------------------------------------------"

hist(seu$nCount_RNA)
# equivalent to
hist(seu@meta.data$nCount_RNA)

# Suerat has built-in plot functions (wraps ggplot2)
Seurat::FeatureScatter(seu, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

"-------------------------------------------------------------------------------
While generating the Seurat object, 
there were already some quality measures calculated for each cell, 
namely the total UMI counts per cell (nCount_RNA) 
and the total number of detected features per cell (nFeature_RNA). 
We can plot those in a violin plot and evaluate their distribution per sample:
-------------------------------------------------------------------------------"
Seurat::VlnPlot(seu, features = c("nCount_RNA",
                                  "nFeature_RNA"))
# A cell with low number of detected features or counts might not give you a lot of information,
# while a high number of detected features/counts might point to doublets.

## TO REMOVE OR NOT TO REMOVE

# Mitochondrial genes:
# A high relative amount of mitochondrial counts can point to damaged cells.
seu <- Seurat::PercentageFeatureSet(seu, 
                                    pattern = "^MT-", # starting with MT
                                    col.name = "percent.mito")

# Ribosomal genes: (NOT rRNA)!
# it can be good to have a look at their relative abundance.
seu <- Seurat::PercentageFeatureSet(seu, 
                                    pattern = "^RP[SL]", 
                                    col.name = "percent.ribo")
# Hemoglobin genes:
# you can expect ‘contamination’ of erythrocytes and select against it.
seu <- Seurat::PercentageFeatureSet(seu,
                                    pattern = "^HB[^(P)]",
                                    col.name = "percent.globin")

# we added there new columns to metadata
head(seu@meta.data) 
# So, the function PercentageFeatureSet adds a column to meta.data,
# specifying the percentage of counts for the specified gene sets.

Seurat::VlnPlot(seu,
                features = c(
                  "percent.mito",
                  "percent.ribo",
                  "percent.globin"))
# PBMMC-2 look different than others
#high percent.globin and low percent.ribo

Seurat::FeatureScatter(seu, 
                       feature1 = "percent.globin", 
                       feature2 = "percent.ribo")

# Check the most expressed------------------------------------------------------
library(ggplot2)
library(Matrix)
library(Seurat)

most_expressed_boxplot <- function(object, ngenes = 20){
  
  # matrix of raw counts
  cts <- Seurat::GetAssayData(object, assay = "RNA", layer = "counts")
  
  # get percentage/cell
  cts <- t(cts)/colSums(cts)*100
  medians <- apply(cts, 2, median)
  
  # get top n genes
  most_expressed <- order(medians, decreasing = T)[ngenes:1]
  most_exp_matrix <- as.matrix((cts[,most_expressed]))
  
  # prepare for plotting
  most_exp_df <- stack(as.data.frame(most_exp_matrix))
  colnames(most_exp_df) <- c("perc_total", "gene")
  
  # boxplot with ggplot2
  boxplot <- ggplot(most_exp_df, aes(x=gene, y=perc_total)) +
    geom_boxplot() +
    coord_flip()
  return(boxplot)
}

most_expressed_boxplot(seu, 20)
# As for most 10X based poly-A enriched single cell datasets, we find a relatively high expression of MALAT1
# ------------------------------------------------------------------------------
# CELL FILTERING
# a threshold of < 8% mitochondrial counts and > 200 features per cell
# To filter against possible doublets, here, we also filter out cells with > 5000 detected features/cell

# Seurat::subset to filter out cells
seu <- subset(seu, 
              subset = nFeature_RNA > 200 &  # not enough signal
                       nFeature_RNA < 5000 & # doublets
                       percent.mito < 8)     # non-intact cells
# to evaluate the filtering, we can visualize
Seurat::VlnPlot(seu, features = c("nFeature_RNA",
                                  "percent.mito"))

## END OF CHAPTER
saveRDS(seu, "seu_after_QC.rds")
